### PR TITLE
Add circular convolution functionality

### PR DESCRIPTION
**Pull request by devashisht**
Summary:
Added the functionality of circular convolution to the
core aten and exposed it through nn.yaml. The functionality
hasn't been exposed to the python API yet.

Test Plan:
pytest test/test_nn.py -k conv

### CODE REVIEW

**Comment by apaszke** [comment]
Hmm it looks like this modified some slow codepaths that no one really should use, and circular convolution can be easily implemented in a few lines of Python. Why can't we simply use the wrappers we have?

**Comment by ezyang** [comment]
@apaszke context is in this issue: #6124

**Comment by apaszke** [comment]
Yeah I know what circular convolution is, my point is that no one should really use that CPU code that computes the Toeplitz matrix. It's extremely memory hungry and very slow compared to cuDNN/MKLDNN. What we should do instead would be a native function which computes something like this:
```
def circular_conv(x, *args):
  end_slice_of_x = x[-kernel_size:] # not sure if should be kernel_size, but probably some function of it
  begin_slice_of_x = x[:kernel_size]
  return regular_conv(torch.cat([end_slice_of_x, x, begin_slice_of_x]), *args)
```

**Comment by apaszke** [comment]
@SsnL I'm pretty sure this can still be done using *an extra padding step*, and then a single convolution. I'd be very surprised if the overhead of padding would be higher than the cost of routing circular convolution through TH(CU)NN conv kernels, which are terrible.

File: aten/src/ATen/native/Convolution.cpp [hunk 1]
```diff
@@ -226,52 +226,52 @@ at::Tensor conv1d(
     const Tensor& input, const Tensor& weight, const Tensor& bias,
     IntList stride, IntList padding, IntList dilation, int64_t groups) {
   return at::convolution(input, weight, bias, stride, padding, dilation,
-                         false, {0}, groups);
+                         false, {0}, groups, false);
```
[the remark this answers is not in the document]
Comment (SsnL) [reply aten/src/ATen/native/Convolution.cpp]: I would prefer this (and other places) to use an enum rather than a boolean :)

**Comment by SsnL** [comment]
@apaszke I agree that THNN code is slow. But if we were to provide such padding options directly on conv, it should still be:
```
if cudnn/mkldnn:
  pad then conv
else:
  # use thnn
  call thnn conv with padding option that fuses the padding into the conv kernel
```

So this is not all pointless. Yeah, for cudnn/mkldnn, we probably want extra padding step. That sounds worth benchmarking.
